### PR TITLE
Add recursive membership sync example

### PR DESCRIPTION
**Pull request by enj**
I have confirmed this works with a real Microsoft Active Directory server.  @stevekuznetsov @liggitt PTAL.  Once this is merged I will ask QA to test.

xref:

https://bugzilla.redhat.com/show_bug.cgi?id=1437324
https://github.com/openshift/origin/issues/12168

Signed-off-by: Monis Khan <mkhan@redhat.com>

### CODE REVIEW

File: install_config/syncing_groups_with_ldap.adoc [hunk 1]
```diff
@@ -229,7 +229,7 @@ $ oadm groups prune --sync-config=config.yaml --confirm
 ----
 
 [[sync-examples]]
-== Sync Examples
+== Direct Membership Sync Examples
```
Comment (stevekuznetsov) [inline install_config/syncing_groups_with_ldap.adoc]: A minority of people need nested groups. Don't make everyone parse this out. Let the people who need nested groups read your note and click your link.

File: install_config/syncing_groups_with_ldap.adoc [hunk 2]
```diff
@@ -653,12 +660,12 @@ cn: Jim
...
 mail: jim.adams@example.com
-testMemberOf: admins
+memberOf: admins
 ----
 <1> The user's group memberships are listed as attributes on the user, and the
-group does not exist as an entry on the server. The `*testMemberOf*` attribute
-cannot be a literal attribute on the user; it can be created during search and
-returned to the client, but not committed to the database.
+group does not exist as an entry on the server. The `*memberOf*` attribute does
+not have to be a literal attribute on the user; it can be created during search
```
Comment (stevekuznetsov) [inline install_config/syncing_groups_with_ldap.adoc]: Maybe `s/it can be/in some LDAP servers, it is/` ?

File: install_config/syncing_groups_with_ldap.adoc [hunk 3]
```diff
@@ -860,3 +867,173 @@ users: <5>
...
+augmented Active Directory with recursive members schema requires you to
+provide an LDAP query definition for user entries and the attributes
+with which to represent them in the internal {product-title} Group records.
+Furthermore, the `oadm groups sync` command must explicitly
+xref:../install_config/syncing_groups_with_ldap.adoc#running-ldap-sync[whitelist]
+groups and the user's `groupMembershipAttributes` must specify
+https://msdn.microsoft.com/en-us/library/aa746475(v=vs.85).aspx[LDAP_MATCHING_RULE_IN_CHAIN].
+`groupUIDAttribute` must be set to `dn` and the `groupsQuery` must not set `filter`.
+`groupsQuery` must set a valid `derefAliases` and should not set `baseDN` or `scope`
+as those values are ignored.
```
Comment (stevekuznetsov) [inline install_config/syncing_groups_with_ldap.adoc]: I think my head exploded ... can we make this a list or something?

File: install_config/syncing_groups_with_ldap.adoc [hunk 4]
```diff
@@ -860,3 +867,173 @@ users: <5>
...
+
+----
+$ oadm groups sync 'cn=admins,ou=groups,dc=example,dc=com' --sync-config=augmented_active_directory_config_recursive.yaml --confirm
+----
+
+[NOTE]
+====
+You must explicitly
+xref:../install_config/syncing_groups_with_ldap.adoc#running-ldap-sync[whitelist]
+the `cn=admins,ou=groups,dc=example,dc=com` group.
```
Comment (stevekuznetsov) [inline install_config/syncing_groups_with_ldap.adoc]: Put the whitelist in the example above

**Review by stevekuznetsov — CHANGES_REQUESTED** [review]

File: install_config/syncing_groups_with_ldap.adoc (hunk 1 above)
Comment (enj) [inline install_config/syncing_groups_with_ldap.adoc]: @liggitt specifically asked for this.

File: install_config/syncing_groups_with_ldap.adoc (hunk 3 above)
Comment (enj) [inline install_config/syncing_groups_with_ldap.adoc]: They are listed out in the example config below this...

File: install_config/syncing_groups_with_ldap.adoc (hunk 4 above)
Comment (enj) [inline install_config/syncing_groups_with_ldap.adoc]: It is in the example above...

File: install_config/syncing_groups_with_ldap.adoc (hunk 1 above)
Comment (stevekuznetsov) [inline install_config/syncing_groups_with_ldap.adoc]: Weird.

File: install_config/syncing_groups_with_ldap.adoc (hunk 3 above)
Comment (stevekuznetsov) [inline install_config/syncing_groups_with_ldap.adoc]: Right, but these are not examples, they are imperative statements. You can't have an example that tells me `baseDN` is ignored.

Comment (stevekuznetsov) [inline install_config/syncing_groups_with_ldap.adoc]: I'm just saying here the sentence/paragraph as is is huge and hard to hold in your mind. Would be easier to consume broken up.

File: install_config/syncing_groups_with_ldap.adoc (hunk 4 above)
Comment (stevekuznetsov) [inline install_config/syncing_groups_with_ldap.adoc]: I am blind

**Comment by enj** [comment]
@stevekuznetsov PTAL.

File: install_config/syncing_groups_with_ldap.adoc (hunk 1 above)
Comment (stevekuznetsov) [inline install_config/syncing_groups_with_ldap.adoc]: Still think changing the name of the core example section just to differentiate it more from a small example at the end of the doc is not the best.

Comment (stevekuznetsov) [inline install_config/syncing_groups_with_ldap.adoc]: This makes sense in the context of "recursive membership" but most people won't even see that.

Comment (enj) [inline install_config/syncing_groups_with_ldap.adoc]: Except that doing crazy group nesting is common with AD which is probably the most used LDAP server...

Comment (enj) [inline install_config/syncing_groups_with_ldap.adoc]: I do not have a strong opinion on this, will let @liggitt decide.

**Comment by enj** [comment]
After some serious hacks I generated the spec for the config struct: d59ac471c300ac3b6b28fd97e9cbe99adfb1c307

File: install_config/syncing_groups_with_ldap.adoc [hunk 5]
```diff
@@ -241,6 +241,13 @@ members: *Jane* and *Jim*. Each example explains:
 * What the LDAP sync configuration file looks like.
 * What the resulting Group record in {product-title} will be after synchronization.
 
+[NOTE]
+====
+These examples assume that all users are direct members of their respective groups.
+Specially, no groups have other groups as members.  See xref:recursive-sync[Recursive Membership Sync Example]
```
Comment (liggitt) [inline install_config/syncing_groups_with_ldap.adoc]: s/Specially/Specifically/?

File: install_config/syncing_groups_with_ldap.adoc [hunk 6]
```diff
@@ -860,3 +867,277 @@ users: <5>
 <4> The name of the Group as specified by the sync file.
 <5> The users that are members of the Group, named as specified by the sync file.
 ====
+
+[[recursive-sync]]
+== Recursive Membership Sync Example
```
Comment (liggitt) [inline install_config/syncing_groups_with_ldap.adoc]: Might want to call this "Nested Membership" instead

File: install_config/syncing_groups_with_ldap.adoc [hunk 7]
```diff
@@ -860,3 +867,277 @@ users: <5>
...
+
+Groups in OpenShift do not nest.  Thus the LDAP server must flatten group membership before
+the data can be consumed.  Microsoft's Active Directory Server supports this feature via the
+https://msdn.microsoft.com/en-us/library/aa746475(v=vs.85).aspx[LDAP_MATCHING_RULE_IN_CHAIN]
```
Comment (liggitt) [inline install_config/syncing_groups_with_ldap.adoc]: the markup for this looks weird to me... is that correct?

File: install_config/syncing_groups_with_ldap.adoc [hunk 8]
```diff
@@ -860,3 +867,277 @@ users: <5>
...
+<1> The last time this Group was synchronized with the LDAP server, in ISO 6801 format.
+<2> The unique identifier for the group on the LDAP server.
+<3> The IP address and host of the LDAP server where this Group's record is stored.
+<4> The name of the Group as specified by the sync file.
+<5> The users that are members of the Group, named as specified by the sync file.
+Note that the recursive members of the *admins* group are included since the group
+membership was flattened by the Microsoft Active Directory Server.
+====
+
+== LDAP Sync Configuration Specification
```
Comment (stevekuznetsov) [inline install_config/syncing_groups_with_ldap.adoc]: Do we have a way to regenerate this using a script hopefully if it changes?

File: install_config/syncing_groups_with_ldap.adoc [hunk 9]
```diff
@@ -860,3 +867,277 @@ users: <5>
...
+This section has an example for the augmented Active Directory schema which synchronizes a
+group named *admins* that has one user *Jane* and one group *otheradmins* as members.
+The *otheradmins* group has one user member: *Jim*.  This example explains:
+
+* How the group and users are added to the LDAP server.
+* What the LDAP sync configuration file looks like.
+* What the resulting Group record in {product-title} will be after synchronization.
+
+In the augmented Active Directory schema, both users (Jane and Jim) and groups
+exist in the LDAP server as first-class entries, and group membership is stored
```
Comment (liggitt) [inline install_config/syncing_groups_with_ldap.adoc]: exist as LDAP entries

File: install_config/syncing_groups_with_ldap.adoc [hunk 10]
```diff
@@ -860,3 +867,277 @@ users: <5>
...
+in attributes on the user or the group. The following snippet of `ldif` defines the
+users and group for this schema:
```
Comment (liggitt) [inline install_config/syncing_groups_with_ldap.adoc]: users and groups

File: install_config/syncing_groups_with_ldap.adoc [hunk 11]
```diff
@@ -860,3 +867,277 @@ users: <5>
...
+description: Other System Administrators
+memberOf: cn=admins,ou=groups,dc=example,dc=com <1> <3>
+member: cn=Jim,ou=users,dc=example,dc=com
+----
+<1> The user's and group's memberships are listed as attributes on the object.
+<2> The groups are first-class entries on the LDAP server.
+<3> The *otheradmins* group is a member of the *admins* group.
+====
+
+To sync this group, you must first create the configuration file. The augmented
```
Comment (liggitt) [inline install_config/syncing_groups_with_ldap.adoc]: To sync nested groups with Active Directory, you must provide an LDAP query definition...

File: install_config/syncing_groups_with_ldap.adoc [hunk 12]
```diff
@@ -860,3 +867,277 @@ users: <5>
...
+Active Directory with recursive members schema requires you to provide an LDAP
+query definition for both user entries and group entries, as well as the attributes
+with which to represent them in the internal {product-title} Group records.
+Furthermore, certain changes are required in this configuration:
+
+- The `oadm groups sync` command must explicitly
+xref:../install_config/syncing_groups_with_ldap.adoc#running-ldap-sync[whitelist] groups.
+- The user's `groupMembershipAttributes` must specify
```
Comment (liggitt) [inline install_config/syncing_groups_with_ldap.adoc]: be explicit here about the attribute you expect

File: install_config/syncing_groups_with_ldap.adoc [hunk 13]
```diff
@@ -860,3 +867,277 @@ users: <5>
...
+file creates these relationships.
+
+.LDAP Sync Configuration Using Augmented Active Directory Schema With Recursive Members: *_augmented_active_directory_config_recursive.yaml_*
+====
+[source,yaml]
+----
+kind: LDAPSyncConfig
+apiVersion: v1
+url: ldap://LDAP_SERVICE_IP:389
+insecure: true
```
Comment (liggitt) [inline install_config/syncing_groups_with_ldap.adoc]: don't encourage insecure mode in our examples

File: install_config/syncing_groups_with_ldap.adoc [hunk 14]
```diff
@@ -860,3 +867,277 @@ users: <5>
...
+        filter: (objectclass=inetOrgPerson)
+        pageSize: 0
+    userNameAttributes: [ mail ] <4>
+    groupMembershipAttributes: [ "memberOf:1.2.840.113556.1.4.1941:" ] <5>
+----
+<1> `groupsQuery` filters cannot be specified.  The `groupsQuery` base DN and scope values are ignored.
+`groupsQuery` must set a valid `derefAliases`.
+<2> The attribute that uniquely identifies a group on the LDAP server. It must be set to `dn`.
+<3> The attribute to use as the name of the Group.
+<4> The attribute to use as the name of the user in the {product-title} Group record.
```
Comment (liggitt) [inline install_config/syncing_groups_with_ldap.adoc]: Not sure if mail or sAMAccountName would be preferred in most installations... can at least note sAMAccountName as an alternative

File: install_config/syncing_groups_with_ldap.adoc [hunk 15]
```diff
@@ -860,3 +867,277 @@ users: <5>
...
+users: <5>
+- jane.smith@example.com
+- jim.adams@example.com
+----
+<1> The last time this Group was synchronized with the LDAP server, in ISO 6801 format.
+<2> The unique identifier for the group on the LDAP server.
+<3> The IP address and host of the LDAP server where this Group's record is stored.
+<4> The name of the Group as specified by the sync file.
+<5> The users that are members of the Group, named as specified by the sync file.
+Note that the recursive members of the *admins* group are included since the group
```
Comment (liggitt) [inline install_config/syncing_groups_with_ldap.adoc]: Note that members of nested groups are included...

File: install_config/syncing_groups_with_ldap.adoc [hunk 16]
```diff
@@ -860,3 +867,277 @@ users: <5>
...
+|augmentedActiveDirectory|AugmentedActiveDirectoryConfig holds the configuration for extracting data from an LDAP server set up in a fashion similar to that used in Active Directory as described above, with one addition: first-class group entries exist and are used to hold metadata but not group membership|false|<<v1.AugmentedActiveDirectoryConfig>>
+|===
+
+=== v1.StringSource
+:hardbreaks:
+StringSource allows specifying a string inline, or externally via env var or file. When it contains only a string value, it marshals to a simple JSON string.
+
+[options="header"]
+|===
+|Name|Description|Required|Schema
```
Comment (liggitt) [inline install_config/syncing_groups_with_ldap.adoc]: the required columns are not very accurate... I'd drop them

File: install_config/syncing_groups_with_ldap.adoc (hunk 7 above)
Comment (enj) [inline install_config/syncing_groups_with_ldap.adoc]: Looks fine to me:

image

File: install_config/syncing_groups_with_ldap.adoc (hunk 8 above)
Comment (enj) [inline install_config/syncing_groups_with_ldap.adoc]: Sort of, I have a fork that I can use to do most of the work.

**Comment by enj** [comment]
@adellape @ahardin-rh This is ready for you guys.

File: install_config/syncing_groups_with_ldap.adoc [hunk 17]
```diff
@@ -886,12 +882,11 @@ The *otheradmins* group has one user member: *Jim*.  This example explains:
...
-In the augmented Active Directory schema, both users (Jane and Jim) and groups
-exist in the LDAP server as first-class entries, and group membership is stored
-in attributes on the user or the group. The following snippet of `ldif` defines the
-users and group for this schema:
+In the augmented Active Directory schema, both users (Jane and Jim) and groups exist
+as LDAP entries, and group membership is stored in attributes on the user or the group.
+The following snippet of `ldif` defines the users and groups for this schema:
 
-.LDAP Entries Using Augmented Active Directory Schema With Recursive Members: *_augmented_active_directory_recursive.ldif_*
+.LDAP Entries Using Augmented Active Directory Schema With Nested Members: *_augmented_active_directory_nested.ldif_*
```
Comment (stevekuznetsov) [inline install_config/syncing_groups_with_ldap.adoc]: I think you're going to get dinged by the docs team for making your lines too long

File: install_config/syncing_groups_with_ldap.adoc [hunk 18]
```diff
@@ -886,12 +882,11 @@ The *otheradmins* group has one user member: *Jim*.  This example explains:
 * What the LDAP sync configuration file looks like.
 * What the resulting Group record in {product-title} will be after synchronization.
 
-In the augmented Active Directory schema, both users (Jane and Jim) and groups
-exist in the LDAP server as first-class entries, and group membership is stored
-in attributes on the user or the group. The following snippet of `ldif` defines the
-users and group for this schema:
+In the augmented Active Directory schema, both users (Jane and Jim) and groups exist
+as LDAP entries, and group membership is stored in attributes on the user or the group.
```
Comment (stevekuznetsov) [inline install_config/syncing_groups_with_ldap.adoc]: Why the change here re: first-class entries?

File: install_config/syncing_groups_with_ldap.adoc [hunk 19]
```diff
@@ -946,16 +941,16 @@ member: cn=Jim,ou=users,dc=example,dc=com
...
+To sync nested groups with Active Directory, you must provide an LDAP query definition for
+both user entries and group entries, as well as the attributes with which to represent them
+in the internal {product-title} Group records. Furthermore, certain changes are required
+in this configuration:
 
 - The `oadm groups sync` command must explicitly
 xref:../install_config/syncing_groups_with_ldap.adoc#running-ldap-sync[whitelist] groups.
 - The user's `groupMembershipAttributes` must specify
 https://msdn.microsoft.com/en-us/library/aa746475(v=vs.85).aspx[LDAP_MATCHING_RULE_IN_CHAIN].
+Thus it must include `"memberOf:1.2.840.113556.1.4.1941:"`.
```
Comment (stevekuznetsov) [inline install_config/syncing_groups_with_ldap.adoc]:
```
 - The user's `groupMembershipAttributes` must include `"memberOf:1.2.840.113556.1.4.1941:"`
   to comply with https://msdn.microsoft.com/en-us/library/aa746475(v=vs.85).aspx[LDAP_MATCHING_RULE_IN_CHAIN].
```

File: install_config/syncing_groups_with_ldap.adoc [hunk 20]
```diff
@@ -969,14 +964,13 @@ fields. For example, identify the users of a Group by their e-mail,
...
 
-.LDAP Sync Configuration Using Augmented Active Directory Schema With Recursive Members: *_augmented_active_directory_config_recursive.yaml_*
+.LDAP Sync Configuration Using Augmented Active Directory Schema With Nested Members: *_augmented_active_directory_config_nested.yaml_*
 ====
 [source,yaml]
 ----
 kind: LDAPSyncConfig
 apiVersion: v1
 url: ldap://LDAP_SERVICE_IP:389
-insecure: true
```
Comment (stevekuznetsov) [inline install_config/syncing_groups_with_ldap.adoc]: Why drop this versus make it `false`?

Comment (stevekuznetsov) [inline install_config/syncing_groups_with_ldap.adoc]: (either is fine but be consistent)

**Review by stevekuznetsov — CHANGES_REQUESTED** [review]

File: install_config/syncing_groups_with_ldap.adoc (hunk 18 above)
Comment (enj) [inline install_config/syncing_groups_with_ldap.adoc]: Because Jordan asked for it and both are clear to me.

File: install_config/syncing_groups_with_ldap.adoc (hunk 17 above)
Comment (enj) [inline install_config/syncing_groups_with_ldap.adoc]: I will let them decide how they want to update this and the preexisting lines with that issue.

File: install_config/syncing_groups_with_ldap.adoc (hunk 18 above)
Comment (stevekuznetsov) [inline install_config/syncing_groups_with_ldap.adoc]: Change the rest of this in the doc then, I'm pretty sure we use that terminology a lot in the whole thing.

File: install_config/syncing_groups_with_ldap.adoc (hunk 20 above)
Comment (enj) [inline install_config/syncing_groups_with_ldap.adoc]: Because it is not relevant to the example.  I only left the others since they had callouts to them.  Ideally they would reference the CA field, but I would prefer to limit this already overly large diff.

**Comment by enj** [comment]
@adellape @ahardin-rh OK, now I am actually done.

**Review by adellape — COMMENTED** [review]
This seems OK for merge after squashing down to a single commit, and I'll do a follow-up for some minor formatting edits (e.g., the `<<` `>>` xrefs, adding unique anchors). 

BTW, how come the `:hardbreaks:`?

**Comment by enj** [comment]
@adellape squashed.

> BTW, how come the `:hardbreaks:`?

No idea, I used the swagger generation tool to create those tables.  They are similar to the ones in `rest_api/openshift_v1.adoc`.

**Comment by adellape** [comment]
[Add recursive membership LDAP sync example](https://trello.com/c/E5jwLh6X/516-add-recursive-membership-ldap-sync-example)
